src/rust/engine/ui/src/display.rs

```diff
@@ -0,0 +1,307 @@
+extern crate rand;
+extern crate termion;
+extern crate unicode_segmentation;
+
+use std::collections::{HashMap, VecDeque};
+use std::io::Write;
+use std::io::{stdout, Result, Stdout};
+use std::thread;
+use std::time::Duration;
+
+use termion::cursor::DetectCursorPos;
+use termion::raw::IntoRawMode;
+use termion::raw::RawTerminal;
+use termion::{clear, color, cursor};
+use unicode_segmentation::UnicodeSegmentation;
+
+enum Console {
+  Terminal(RawTerminal<Stdout>),
+  Pipe(Stdout),
+}
+
+pub struct EngineDisplay {
+  sigil: char,
+  divider: String,
+  poll_interval_ms: Duration,
+  padding: String,
+  terminal: Console,
+  action_map: HashMap<String, String>,
```

Comment:
Should this be a `BTreeMap`? It looks like when we consume from it, we always want it to be sorted; just storing it sorted is probably cheaper than sorting every time we consume?

Reply:
works for me! fixed